src/java/org/apache/cassandra/index/sai/disk/StorageAttachedIndexWriter.java

```diff
@@ -0,0 +1,255 @@
+                                                                                               rowMapping))
+                                      .filter(Objects::nonNull) // a null here means the column had no data to flush
+                                      .collect(Collectors.toList());
+
+        // If the SSTable components are already being built by another index build then we don't want
+        // to build them again so use a null writer
+        this.perSSTableWriter = perIndexComponentsOnly ? PerSSTableWriter.NONE : indexDescriptor.newPerSSTableWriter();
+    }
+
+    @Override
+    public void begin()
+    {
+        logger.debug(indexDescriptor.logMessage("Starting partition iteration for storage attached index flush for SSTable {}..."), indexDescriptor.descriptor);
+        stopwatch.start();
+    }
+
+    @Override
+    public void startPartition(DecoratedKey key, long position)
+    {
+        if (aborted) return;
+        
+        currentKey = key;
+    }
+
+    @Override
+    public void nextUnfilteredCluster(Unfiltered unfiltered)
+    {
+        if (aborted) return;
+
+        // Ignore range tombstones...
+        if (!unfiltered.isRow())
+            return;
+
+        try
+        {
+            addRow((Row)unfiltered);
+        }
+        catch (Throwable t)
+        {
+            logger.error(indexDescriptor.logMessage("Failed to record a row during an index build"), t);
+            abort(t, true);
+        }
+    }
+
+    @Override
+    public void partitionLevelDeletion(DeletionTime deletionTime, long position)
+    {
+        // Deletions (including partition deletions) are accounted for during reads.
+    }
+
+    @Override
+    public void staticRow(Row staticRow, long position)
+    {
+        if (aborted) return;
+        
+        if (staticRow.isEmpty())
+            return;
+
+        try
+        {
+            addRow(staticRow);
+        }
+        catch (Throwable t)
+        {
+            logger.error(indexDescriptor.logMessage("Failed to record a static row during an index build"), t);
+            abort(t, true);
+        }
+    }
+
+    @Override
+    public void complete()
+    {
+        if (aborted) return;
+
+        long start = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+        logger.debug(indexDescriptor.logMessage("Completed partition iteration for index flush for SSTable {}. Elapsed time: {} ms"),
+                     indexDescriptor.descriptor,
+                     start);
+
+        try
+        {
+            perSSTableWriter.complete();
+            tokenOffsetWriterCompleted = true;
+
+            long elapsed = stopwatch.elapsed(TimeUnit.MILLISECONDS);
+
+            logger.debug(indexDescriptor.logMessage("Completed per-SSTable write for SSTable {}. Duration: {} ms. Total elapsed time: {} ms."),
+                         indexDescriptor.descriptor,
+                         elapsed - start,
+                         elapsed);
```

Comment:
nit: Here and on 187, might as well flatten the substitution values into a single line after the string.